engine/execution/ingestion/engine.go

```diff
@@ -376,182 +342,242 @@ func (e *Engine) executeBlock(ctx context.Context, executableBlock *entity.Execu
+			// find the block that was just executed
+			executionQueue, exists := executionQueues.ByID(executed.ID())
 			if !exists {
 				return fmt.Errorf("fatal error - executed block not present in execution queue")
 			}
 
+			// dismount the executed block and all its children
 			_, newQueues := executionQueue.Dismount()
 
+			// go through each children, add them back to the queue, and check
+			// if the children is executable
 			for _, queue := range newQueues {
-				newExecutableBlock := queue.Head.Item.(*entity.ExecutableBlock)
-				newExecutableBlock.StartState = finalState
+				added := executionQueues.Add(queue)
+				if !added {
+					return fmt.Errorf("fatal error - child block already in execution queue")
+				}
+
+				// the parent block has been executed, update the StartState of
+				// each child block.
+				child := queue.Head.Item.(*entity.ExecutableBlock)
+				child.StartState = finalState
 
-				err := e.matchOrRequestCollections(newExecutableBlock, blockByCollection)
+				err := e.matchOrRequestCollections(child, blockByCollection)
 				if err != nil {
 					return fmt.Errorf("cannot send collection requests: %w", err)
 				}
 
-				added := executionQueues.Add(queue)
-				if !added {
-					return fmt.Errorf("fatal error - child block already in execution queue")
+				completed := e.executeBlockIfComplete(child)
+				if !completed {
+					e.log.Debug().
+						Hex("executed_block", logging.Entity(executed)).
+						Hex("child_block", logging.Entity(child)).
+						Msg("child block is not ready to be executed yet")
+				} else {
+					e.log.Debug().
+						Hex("executed_block", logging.Entity(executed)).
+						Hex("child_block", logging.Entity(child)).
+						Msg("child block is ready to be executed")
 				}
-
-				e.executeBlockIfComplete(newExecutableBlock)
 			}
 
-			executionQueues.Rem(executableBlock.ID())
+			// remove the executed block
+			executionQueues.Rem(executed.ID())
 
 			return nil
 		})
 
 	if err != nil {
 		e.log.Err(err).
-			Hex("block_id", logging.Entity(executableBlock)).
+			Hex("block", logging.Entity(executed)).
 			Msg("error while requeueing blocks after execution")
 	}
 
-	e.log.Info().
-		Hex("block_id", logging.Entity(executableBlock)).
-		Uint64("block_height", executableBlock.Block.Header.Height).
-		Hex("final_state", finalState).
-		Msg("block executed")
-	e.metrics.ExecutionLastExecutedBlockHeight(executableBlock.Block.Header.Height)
+	return nil
 }
 
+// executeBlockIfComplete checks whether the block is ready to be executed.
+// if yes, execute the block
+// return a bool indicates whether the block was completed
 func (e *Engine) executeBlockIfComplete(eb *entity.ExecutableBlock) bool {
+	if !eb.HasStartState() {
+		return false
+	}
+
+	// if the eb has parent statecommitment, and we have the delta for this block
+	// then apply the delta
+	// note the block ID is the delta's ID
+	delta, found := e.syncDeltas.ByID(eb.Block.ID())
```

Comment:
```suggestion
	delta, found := e.syncDeltas.ByBlockID(eb.Block.ID())
```

Thanks for the clarifying comment, I'd suggest changing the method name as well.